File: go/problems/reverse_only_letters.go
package problems

// https://leetcode.com/problems/reverse-only-letters/
func ReverseOnlyLetters(S string) string {
  res := []byte(S)
  i, j := 0, len(S)-1
  for i <= j {
    switch {
    case isLetter(res[i]) && isLetter(res[j]):
      res[i], res[j] = res[j], res[i]
      i++
      j--
    case isLetter(res[i]):
      j--
    case isLetter(res[j]):
      i++
    default:
      i++
      j--
    }
  }
  return string(res)
}

func isLetter(i uint8) bool {
  if (i >= 'a' && i <= 'z') || (i >= 'A' && i <= 'Z') {
    return true
  }
  return false
}

File: rust/src/problems/reverse_only_letters.rs
// https://leetcode.com/problems/reverse-only-letters/
pub fn reverse_only_letters(s: String) -> String {
  let mut end: usize = s.len() - 1;
  let reversed: String = s
    .chars()
    .map(|letter| {
      let mut tmp: char = letter;
      if letter.is_alphabetic() {
        while !s.chars().nth(end).unwrap().is_alphabetic() {
          end -= 1;
        }
        tmp = s.chars().nth(end).unwrap();
        if end != 0 {
          end -= 1;
        }
      }
      return tmp;
    })
    .collect::<String>();
  return reversed;
}